src/test/tracknumberstest.cpp

```diff
@@ -0,0 +1,137 @@
+        EXPECT_TRUE(actualTrackNumbers.isValid());
+        EXPECT_EQ(expectedResult, actualTrackNumbers);
+
+        return actualTrackNumbers;
+    }
+
+    TrackNumbers parseFromStringInvalid(const QString& inputValue, const TrackNumbers& expectedResult) {
+        TrackNumbers actualTrackNumbers;
+        TrackNumbers::ParseResult actualParseResult = TrackNumbers::parseFromString(inputValue, &actualTrackNumbers);
+
+        qDebug() << "parseFromString():" << inputValue << "->" << actualTrackNumbers.toString();
+
+        EXPECT_EQ(TrackNumbers::ParseResult::INVALID, actualParseResult);
+        EXPECT_EQ(expectedResult, actualTrackNumbers);
+
+        return actualTrackNumbers;
+    }
+
+    void toString(const TrackNumbers& inputValue, const QString& expectedResult) {
+        const QString actualResult(inputValue.toString());
+        qDebug() << "toString():" << expectedResult << "->" << actualResult;
+
+        EXPECT_EQ(expectedResult, actualResult);
+
+        // Re-parse the formatted string if the input is
+        // a valid TrackNumbers instance
+        if (inputValue.isValid()) {
+            if (actualResult.isEmpty()) {
+                parseFromStringEmpty(actualResult);
+            } else {
+                parseFromStringValid(actualResult, inputValue);
+            }
+        }
+    }
+
+    void splitAndJointStrings(const QString& inputValue, const QString& expectedTrackNumber, const QString& expectedTrackTotal) {
+        QString actualTrackNumber;
+        QString actualTrackTotal;
+        TrackNumbers::splitString(inputValue, &actualTrackNumber, &actualTrackTotal);
+        qDebug() << "splitString():" << inputValue << "->" << actualTrackNumber << actualTrackTotal;
+
+        EXPECT_EQ(expectedTrackNumber, actualTrackNumber);
+        EXPECT_EQ(expectedTrackTotal, actualTrackTotal);
+        EXPECT_EQ(inputValue, TrackNumbers::joinStrings(actualTrackNumber, actualTrackTotal));
+    }
+};
+
+TEST_F(TrackNumbersTest, ParseFromStringEmpty) {
+    parseFromStringEmpty("");
+    parseFromStringEmpty(" \t \n   ");
+}
+
+TEST_F(TrackNumbersTest, ParseFromStringValid) {
+    parseFromStringValid("0", TrackNumbers(0));
+    parseFromStringValid("0/0", TrackNumbers(0, 0));
+    parseFromStringValid("1 / 0", TrackNumbers(1, 0));
+    parseFromStringValid(" 1\t/\t2  ", TrackNumbers(1, 2));
+    parseFromStringValid("12/7", TrackNumbers(12, 7));
+    parseFromStringValid("01234/12345", TrackNumbers(1234, 12345));
+}
+
+TEST_F(TrackNumbersTest, ParseFromStringInvalid) {
+    parseFromStringInvalid("-2", TrackNumbers(-2));
+    parseFromStringInvalid("1/ -1", TrackNumbers(1, -1));
+    parseFromStringInvalid("-2 / 1", TrackNumbers(-2, 1));
+    parseFromStringInvalid("-3/-4", TrackNumbers(-3, -4));
+    parseFromStringInvalid("1/a", TrackNumbers(1, TrackNumbers::kValueUndefined));
+    parseFromStringInvalid("a /1", TrackNumbers(TrackNumbers::kValueUndefined, 1));
```

Comment:
Thanks, Daniel. After adding more test cases I found another minor 
inconsistency in the parsing algorithm ;)

On 01/04/2016 08:35 AM, Daniel Schürmann wrote:
>
> In src/test/tracknumberstest.cpp 
> <https://github.com/mixxxdj/mixxx/pull/835#discussion_r48708987>:
>
> > +TEST_F(TrackNumbersTest, ParseFromStringValid) {
> > +    parseFromStringValid("0", TrackNumbers(0));
> > +    parseFromStringValid("0/0", TrackNumbers(0, 0));
> > +    parseFromStringValid("1 / 0", TrackNumbers(1, 0));
> > +    parseFromStringValid(" 1\t/\t2  ", TrackNumbers(1, 2));
> > +    parseFromStringValid("12/7", TrackNumbers(12, 7));
> > +    parseFromStringValid("01234/12345", TrackNumbers(1234, 12345));
> > +}
> > +
> > +TEST_F(TrackNumbersTest, ParseFromStringInvalid) {
> > +    parseFromStringInvalid("-2", TrackNumbers(-2));
> > +    parseFromStringInvalid("1/ -1", TrackNumbers(1, -1));
> > +    parseFromStringInvalid("-2 / 1", TrackNumbers(-2, 1));
> > +    parseFromStringInvalid("-3/-4", TrackNumbers(-3, -4));
> > +    parseFromStringInvalid("1/a", TrackNumbers(1, TrackNumbers::kValueUndefined));
> > +    parseFromStringInvalid("a /1", TrackNumbers(TrackNumbers::kValueUndefined, 1));
>
> What should happen if a number is followed by rubbish:
> "2 Tracks" or " "1 of 12" I have not seen this in the wild, but we should 
> clarify by a test if we read it or not.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/835/files#r48708987>.
>